File: src/main/java/cube8540/oauth/authentication/users/application/DefaultUserPasswordService.java
package cube8540.oauth.authentication.users.application;

import cube8540.oauth.authentication.users.domain.User;
import cube8540.oauth.authentication.users.domain.UserCredentialsKeyGenerator;
import cube8540.oauth.authentication.users.domain.UserKeyMatchedResult;
import cube8540.oauth.authentication.users.domain.UserRepository;
import cube8540.oauth.authentication.users.domain.UserValidatorFactory;
import cube8540.oauth.authentication.users.domain.Username;
import cube8540.oauth.authentication.users.domain.exception.UserNotFoundException;
import cube8540.oauth.authentication.users.infra.DefaultUserCredentialsKeyGenerator;
import lombok.Setter;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.security.Principal;

@Service
public class DefaultUserPasswordService implements UserPasswordService {

    private final UserRepository repository;

    private final PasswordEncoder encoder;

    @Setter
    private UserCredentialsKeyGenerator keyGenerator = new DefaultUserCredentialsKeyGenerator();

    @Setter(onMethod_ = {@Autowired, @Qualifier("defaultUserValidatorFactory")})
    private UserValidatorFactory validatorFactory;

    @Autowired
    public DefaultUserPasswordService(UserRepository repository, PasswordEncoder encoder) {
        this.repository = repository;
        this.encoder = encoder;
    }

    @Override
    @Transactional
    public UserProfile changePassword(Principal principal, ChangePasswordRequest changeRequest) {
        User user = getUser(principal.getName());

        user.changePassword(changeRequest.getExistingPassword(), changeRequest.getNewPassword(), encoder);
        user.validation(validatorFactory);
        user.encrypted(encoder);

        return UserProfile.of(repository.save(user));
    }

    @Override
    @Transactional
    public UserProfile forgotPassword(String username) {
        User user = getUser(username);
        user.forgotPassword(keyGenerator);
        return UserProfile.of(repository.save(user));
    }

    @Override
    public boolean validateCredentialsKey(String username, String credentialsKey) {
        User user = getUser(username);

        return user.getPasswordCredentialsKey().matches(credentialsKey).equals(UserKeyMatchedResult.MATCHED);
    }

    @Override
    @Transactional
    public UserProfile resetPassword(ResetPasswordRequest resetRequest) {
        User user = getUser(resetRequest.getUsername());

        user.resetPassword(resetRequest.getCredentialsKey(), resetRequest.getNewPassword());
        user.validation(validatorFactory);
        user.encrypted(encoder);

        return UserProfile.of(repository.save(user));
    }

    private User getUser(String username) {
        return repository.findById(new Username(username))
                .orElseThrow(() -> UserNotFoundException.instance(username + " is not found"));
    }
}

File: src/main/kotlin/cube8540/oauth/authentication/users/application/DefaultUserPasswordService.kt
package cube8540.oauth.authentication.users.application

import cube8540.oauth.authentication.users.domain.User
import cube8540.oauth.authentication.users.domain.UserCredentialsKeyGenerator
import cube8540.oauth.authentication.users.domain.UserKeyMatchedResult
import cube8540.oauth.authentication.users.domain.UserNotFoundException
import cube8540.oauth.authentication.users.domain.UserRepository
import cube8540.oauth.authentication.users.domain.UserValidatorFactory
import cube8540.oauth.authentication.users.domain.Username
import cube8540.oauth.authentication.users.infra.DefaultUserCredentialsKeyGenerator
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class DefaultUserPasswordService @Autowired constructor(
    private val repository: UserRepository,
    private val encoder: PasswordEncoder
): UserPasswordService {

    @set:[Autowired Qualifier("userPasswordChangeValidatorFactory")]
    lateinit var validatorFactory: UserValidatorFactory

    var keyGenerator: UserCredentialsKeyGenerator = DefaultUserCredentialsKeyGenerator()

    @Transactional
    override fun changePassword(username: String, changeRequest: ChangePasswordRequest): UserProfile {
        val user = getUser(username)

        user.changePassword(changeRequest.existingPassword, changeRequest.newPassword, encoder)
        user.validation(validatorFactory)
        user.encrypted(encoder)

        return UserProfile(repository.save(user))
    }

    @Transactional
    override fun forgotPassword(username: String): UserProfile {
        val user = getUser(username)

        user.forgotPassword(keyGenerator)
        return UserProfile(repository.save(user))
    }

    @Transactional
    override fun validateCredentialsKey(username: String, credentialsKey: String): Boolean {
        val user = getUser(username)

        return user.passwordCredentialsKey?.matches(credentialsKey) == UserKeyMatchedResult.MATCHED
    }

    @Transactional
    override fun resetPassword(resetRequest: ResetPasswordRequest): UserProfile {
        val user = getUser(resetRequest.username)

        user.resetPassword(resetRequest.credentialsKey, resetRequest.newPassword)
        user.validation(validatorFactory)
        user.encrypted(encoder)

        return UserProfile(repository.save(user))
    }

    private fun getUser(username: String): User = repository
        .findById(Username(username))
        .orElseThrow { UserNotFoundException.instance("$username is not found") }
}